### PR TITLE
Simplify TransportNodesAction

### PR DESCRIPTION
**Pull request by DaveCTurner**
Recent improvements to the primitives for writing async code (particularly #92452 and #92620) mean that we can enormously simplify `TransportNodesAction`. In particular, we can avoid accumulating an intermediate array of responses for later processing in favour of just accumulating the successes and failures into their final separate lists. We also no longer need to use a separate `NodesResponseTracker` to discard responses on cancellation.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
@gmarouli I thought you'd be interested in this because of your work in this area, particularly #82685.

**Comment by gmarouli** [comment]
@DaveCTurner thanks! Indeed, I am :). I am catching up on the changes and I will review it right away!

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java [hunk 1]
```diff
@@ -123,40 +126,90 @@ protected TransportNodesAction(
...
-     *
-     * @param request The associated request.
-     * @param nodeResponseTracker All node-level responses collected so far
-     * @throws NodeResponseTracker.DiscardedResponsesException if {@code nodeResponseTracker} has already discarded the intermediate results
-     * @see #newResponseAsync(Task, BaseNodesRequest, List, List, ActionListener)
-     */
-    // exposed for tests
-    void newResponse(Task task, NodesRequest request, NodeResponseTracker nodeResponseTracker, ActionListener<NodesResponse> listener)
-        throws NodeResponseTracker.DiscardedResponsesException {
+        final var mutex = new Object();
```
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:134]: I am doubting if we need the extra mutex here. I understand accessing the `responses` and to the `exceptions` individually needs to be synchronized but does accessing either of them need to be synchronized too?
If what I am saying above is not clear, maybe the following example will explain better what I have in mind. I am wondering if using the respective list as a lock is enough. For example:

```
              final ActionListener<NodeResponse> nodeResponseListener = ActionListener.notifyOnce(new ActionListener<>() {
                    @Override
                    public void onResponse(NodeResponse nodeResponse) {
                        synchronized (responses) {
                            responses.add(nodeResponse);
                        }
                    }

                    @Override
                    public void onFailure(Exception e) {
                        if (task instanceof CancellableTask cancellableTask && cancellableTask.isCancelled()) {
                            return;
                        }

                        logger.debug(() -> format("failed to execute [%s] on node [%s]", actionName, node), e);
                        synchronized (exceptions) {
                            exceptions.add(new FailedNodeException(node.getId(), "Failed node [" + node.getId() + "]", e));
                        }
                    }
```

What are your thoughts on this? Am I missing something?

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java [hunk 2]
```diff
@@ -123,40 +126,90 @@ protected TransportNodesAction(
...
+                if (cancellableTask.notifyIfCancelled(resultListener)) {
+                    return;
+                }
+            }
+            // ref releases all happen-before here so no need to be synchronized
+            threadPool.executor(finalExecutor)
+                .execute(ActionRunnable.wrap(resultListener, l -> newResponseAsync(task, request, responses, exceptions, l)));
+        });
+
+        final var nodeFailureListeners = new ListenableFuture<NodeResponse>();
```
Comment (gmarouli) [inline server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:150]: If I am not mistaken this is mainly used to notify about the task being cancelled right? What do you think about naming it `nodeCancellationListener` or something similar. My reasoning is that we use quite a few listeners here and it was a bit hard for me to follow.

**Review by gmarouli — COMMENTED** [review]
Nice work, I have some minor comments and a few questions that will help me learn too. Thank you for improving this.

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java [hunk 3]
```diff
@@ -85,7 +88,7 @@ protected TransportNodesAction(
         this.nodeResponseClass = Objects.requireNonNull(nodeResponseClass);
 
         this.transportNodeAction = actionName + "[n]";
-        this.finalExecutor = finalExecutor;
+        this.finalExecutor = finalExecutor.equals(ThreadPool.Names.SAME) ? ThreadPool.Names.GENERIC : finalExecutor;
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:91]: That's right - this was the behaviour before today too:

https://github.com/elastic/elasticsearch/blob/a486a6a6b9fe293585911a6b7367e9abc932c188/server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java#L303

(IMO this is kinda questionable, but that's another debate)

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:150]: Yes that's right this is just used to complete the outstanding node listeners (releasing refs to `responses` and `exceptions`) on cancellation. Naming things is hard :)

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:91]: Ok, I am curious about that debate out of the scope of this PR.  I might poke you at some point if you are available.

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java (hunk 2 above)
Comment (gmarouli) [reply server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java]: Indeed, it is :)

**Review by gmarouli — APPROVED** [review]
I didn't get a ping to review again but all my comments were addressed so LGTM 🚀 
Thanks @DaveCTurner

**Comment by DaveCTurner** [comment]
Thanks @gmarouli !

**Pull request merged**